File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/auth/application/service/PasswordHasher.java
package com.example.accountservice.auth.application.service;

/**
 * A Technical Service interface for password hashing (see domain-service.md). It is defined in the
 * shape the caller (the Application layer) needs, and does not expose details of the actual hashing
 * algorithm (BCrypt, etc.). The implementation lives in the infrastructure/ layer
 * (BCryptPasswordHasher).
 */
public interface PasswordHasher {

    String hash(String plainPassword);

    boolean verify(String plainPassword, String passwordHash);
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/auth/application/service/PasswordHasher.kt
package com.example.accountservice.auth.application.service

/**
 * The password hashing/verification Technical Service — the interface lives in the Application layer
 * (in the shape the caller needs), and the implementation (BCrypt) lives in infrastructure/
 * (the Technical Service pattern from domain-service.md, the same structure as
 * secret/application/service/SecretService.kt).
 */
interface PasswordHasher {
    fun hash(plainPassword: String): String

    fun verify(
        plainPassword: String,
        passwordHash: String,
    ): Boolean
}
